advanced/src/hashTable.js

```diff
@@ -0,0 +1,199 @@
+    this.length = 0
+    this.head = null
+  }
+
+  insert(key, value) {
+    let node = new Node(key, value)
+    let currentNode = this.head
+
+    if (!currentNode) {
+      this.head = node
+      this.length++
+      return node
+    }
+    while (currentNode.next) {
+      currentNode = currentNode.next
+    }
+    currentNode.next = node
+    this.length += 1
+    return node
+  }
+
+//regular find
+  find(key) {
+    var currentNode = this.head
+    if (!currentNode) {
+      return null
+    }
+    while (currentNode) {
+      if (currentNode.key === key){
+        return currentNode
+      } else {
+        currentNode = currentNode.next
+      }
+    } return -1
+  }
+
+//finds the node whose next value is the target node
+  findByNext(key) {
+    let currentNode = this.head
+
+    if(!currentNode) {
+      return null
+    } else if (!currentNode.next && currentNode.key === key){
+      return clear()
+    } else {
+      while(currentNode.next){
+        if(currentNode.next.key === key){
+          return currentNode
+        } else {
+          currentNode = currentNode.next
+        }
+      } return null
+    }
+  }
+
+//cleaner remove using findByNext
+  removeWithFindByNext(key) {
+    let currentNode = this.head
+
+    const findByNext = function(key) {
+      if(!currentNode) {
+        return null
+      } else if (!currentNode.next && currentNode.key === key){
+        return clear()
+      } else {
+        while(currentNode.next){
+          if(currentNode.next.key === key){
+            return currentNode
+          } else {
+            currentNode = currentNode.next
+          }
+        } return null
+      }
+    }
+
+    let foundNode = findByNext(key)
+    if (foundNode === null){
+      return null
+    } else {
+      foundNode.next = foundNode.next.next
+      return this.length--
+    }
+  }
+//brute force
+  removeIt(key) {
+    let currentNode = this.head
+
+    if(!currentNode){
+      return null
+    } else {
+      while(currentNode){
+        if(currentNode.key === key){
+          if(!currentNode.next){
+            this.head = null
+            return this.length--
+          } else {
+            currentNode = currentNode.next
+            this.length--
+            return this.head = currentNode
+          }
+        } else if (currentNode.next.key === key){
+          currentNode.next = currentNode.next.next
+          return this.length--
+        } else if (currentNode.next !== null){
+          currentNode = currentNode.next
+        } else {
+          return null
+        }
+      }
+    }
+  }
+
+}
+
+export default class HashTable {
+  constructor() {
+    this.length = 0
+    this.table = {}
+  }
+
+  //ht.hash("name")
+    // generates a hash for the key "name"
+  hash( str ) {
+    const inputString = str
+    const strToArray = inputString.toLowerCase().split('')
+    const alphabit = ['a','b','c','d','e','f','g','h','i','j','k','l','m','n','o','p','q','r','s','t','u','v','w','x','y','z']
```

Comment:
Awesome suggestions, that one-liner is so cool! Now I wanna know what the first 97 CharCodes are! (adds to queue of things to look up)